README.md
Coursework for Distributed Systems module as part of MSc Computer Science degree. The goal was to add some functionality to handle networking and consistency issues in existing multiplayer games. 

Original code forked from github.com/apetenchea/SpaceInvaders

See CHANGES.txt and diff.diff file to see the updates/changes made to the game as below:

The additional functionality required was as follows:

- Introduce an artificial network delay to simulate some of the challenges which take place in networked games.
- Implement bucket synchronization so that all clients have the same frame rate so the game can be considered "fair"

- Implement dead reckoning so that players continue to update in the event that network problems cause delays in updates received about player position
- Implement smooth corrections so that players move towards the correct position after drops in network communication to prevent jerky animation

- Introduce some form of cheat which allows one client to perform a game action that cannot be done by other clients. This can take multiple forms. One example would be to increase the movement speed of one player
- Introduce some checking on the server side which allows the cheating to be detected

- Implement interest management so that a large map exists on the primary copy of the game while information on objects which are sent to secondary copies should only contain information relavent to that client



CHANGES.txt
For a diff file containing all amendments and changes to the code, see diff.diff file in root directory or clone the original repo and run the following Linux command:
diff -ENwbur -x ".git" -x "*.html" -x "*.css" -x "*target*" -x ".vscode*" ./ ../Original/SpaceInvaders/ > diff.diff

Artificial Network delay
    - In client/ClientConfig.java, each player is assigned a random ping with ThreadLocalRandom from a given range
    - In client/network/UdpSender.java:sendPacket() we Thread.sleep() by this ping value before sending each packet, simulating network latency
    - This ping integer is sent to the server along with the other Player information (e.g. username etc) via the SetPlayerIdCommand as the server requires it for bucket synchronisation
    - The server stores the ping with the player in their instance of Player.java

Bucket Synchro
    - Game.java - before we start the game, we compare every players ping, and set a delay for each Player. this is the difference between that players ping and the person with the highest ping. we can then delay execution of events for each player by this amount.
    - GameLoop.java - execution of events is done in GameLoop.java, we loop through each player in the list and execute their commands
        - If we were to delay one players commands with a Thread.Sleep(delay) here, all commands and players after this would also be affected by the delay. For example, if the first player when we loop through the players list had a 50ms delay before execution, the next player to have their commands executed would also have had to wait 50ms
        - To resolve this, in Player.java we have implemented Comparable to be able to sort our player list by ping delay (overriding the compareTo() function)
        - We sort the player list by ping delay (lowest or no delay first), and execute these commands first, and delaying by the appropriate amount
        - We then keep track of how much we've already delayed, so for each successive player we can deduct the delay that has already occurred

    - There is also an implementation of "buckets" similar to Nagles algorithm. The standard code sent an individual UDP packet for each command, rather than grouping multiple commands into one packet to save network traffic.
        - In client/network/UdpSender.java there is handleBucket() which is a recursive function that takes an array of commands and splits it into a single combined JSON packet to be sent over the network. if the JSON output is larger than MAX_PACKET_SIZE, we recursively call the same function on the tail of the command list after sending the first bucket.
            - We delimit JSON commands in the packet using the "~" token
        - In client/mvc/GameModel.java we add commands to the bucket and there is a scheduled executor that sends the command buckets every 250 ms (BUCKET_DELAY var). This limits the rate that clients can send updates to the server, so faster latency players have less of an advantage (and latency is also adjusted for as above).
        - There is corresponding SERVER code to handle multiple commands in one packet in command/CommandBuilder.java, saving incoming commands into an array to be later called and executed from Connection.java
        - To improve on this approach we could handle incoming buckets on the server side in rounds, however this has not currently been implemented.
    - Key files for client side buckets:
        - client/mvc/GameModel.java - adds command to the bucket and sends the bucket client side at specified interval
        - client/network/UdpSender.java - client code to send commands as a JSON packet
        - command/CommandBuilder.java - buildCommand() builds the commandArray from incoming JSON
        - server/network/Connection.java - executeCommandArray() executes the commandArray server side

Smooth Corrections
    - In client/gui/entities/GraphicalEntity.java we have amended relocate() to move to the new position from the old position in 1 pixel intervals, with a 10ms Thread.sleep() inbetween to further smooth out the movement
    - TEST SMOOTH CORRECTIONS
        - Connection.java - TEST_SMOOTH_CORRECTIONS = true
        - GameLoop.java - Set PLAYER_POS_UPDATE to 10000 to delay player updates from server
        - GamePanel.java - Set DEADRECK_ENABLED to false

Dead Reckoning 
    - default settings: DEADRECK_DELAY = 300 ms, PLAYER_POS_UPDATE 150 ms
    - SERVER SIDE: In GameLoop.java we have made a new AutoSwitch called playerPosUpdate that is called from entitiesMove(), this sends an update of player position to the client every 150 ms via a MoveEntityCommand.
    - CLIENT SIDE: in GamePanel.java we have an executor that executes deadReckon() every 300 ms. This gets the previous direction and position (saved in the hashmap's prevEntityDirection amd prevEntityPosition when relocateEntity() is called) and moves the player in that direction.

    - to test and see how ships automatically move in absence of server update
        - GameLoop.java - set PLAYER_POS_UPDATE in GameLoop to 5000 or higher
        - GamePanel.java - set DEADRECK_DEBUG to true to enable debugging messages

Cheat and detection
    - In client/mvc/GameController.java we check if the player name == CHEAT_NAME, and if so set the cheatEnabled flag for the player
    - In the same file, if cheats are enabled, every time the player shoots, 11 bullets entities are created on top of each other. This enables the player bullet to go through multiple invaders/invader bullets.
    - Cheat detection is handled in command/CommandBuilder.java - we detect if there are more than 7 PlayerShootCommand's in the commandArray. If this is detected, we display and message and do not execute any of the commands in this commandArray.

Interest Management
    - Game.java
        - RefreshEntitiesCommand() is a command that updates/pushes every single entity from the game server to the client

src/main/java/spaceinvaders/client/gui/GamePanel.java
package spaceinvaders.client.gui;

import static spaceinvaders.game.EntityEnum.INVADER;
import static spaceinvaders.game.EntityEnum.PLAYER;
import static spaceinvaders.game.EntityEnum.PLAYER_BULLET;

import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.NavigableMap;
import java.util.TreeMap;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import javax.swing.JPanel;
import spaceinvaders.client.gui.entities.Drawable;
import spaceinvaders.client.gui.entities.GraphicalEntity;
import spaceinvaders.client.gui.entities.GraphicalEntityVisitor;
import spaceinvaders.client.gui.entities.GraphicsFactory;
import spaceinvaders.client.gui.entities.PaintingVisitor;
import spaceinvaders.client.gui.entities.Player;
import spaceinvaders.command.server.MovePlayerRightCommand;
import spaceinvaders.game.Entity;
import spaceinvaders.game.EntityEnum;
import spaceinvaders.game.GameConfig;
import spaceinvaders.utility.Couple;

import java.util.HashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Main panel of the game.
 *
 * <p>Contains all the visible elements of the game environment. Controls painting and repainting.
 */
@SuppressWarnings("serial")
class GamePanel extends JPanel {
  private final GraphicsFactory factory = GraphicsFactory.getInstance();
  private final NavigableMap<Integer,GraphicalEntity> entityMap = new TreeMap<>();
  private final ReadWriteLock entitiesLock = new ReentrantReadWriteLock();
  private final SoundManager sound = new SoundManager();
  private final GameConfig config = GameConfig.getInstance();
  private Integer playerAvatarNumber;
  private BufferedImage centerImg;
  private Couple<Integer,Integer> centerImgPos;
  private Boolean gameOn;

  // Dead Reckoning - to test set PLAYER_POS_UPDATE in GameLoop to 5000 and DEADRECK_DELAY below to 500
  private final Boolean DEADRECK_ENABLED = true;
  private final Boolean DEADRECK_DEBUG = false;
  private HashMap<Integer, Integer> prevEntityDirection = new HashMap<Integer, Integer>(); // store the previous direction to calculate dead reck
  private HashMap<Integer, Integer> prevEntityPosition = new HashMap<Integer, Integer>(); // store the previous position (key=entityID, value=x coord)
  // schedule the position of each player to be updated every X ms
  private final int DEADRECK_DELAY = 300; // milliseconds
  ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);

  Runnable deadReckon = new Runnable() {
    public void run() {
        if(!prevEntityDirection.isEmpty()){
          for(Map.Entry<Integer, Integer> entry : prevEntityDirection.entrySet()){
            //System.out.println(prevEntityDirection);
            Integer id = entry.getKey(); 
            Integer direction = entry.getValue();
            final GraphicalEntity entity = entityMap.get(id);
            if (entity == null) {
              throw new NullPointerException();
            }
            if(direction == 0) { // not moved
              if(DEADRECK_DEBUG == true) { System.out.println("Dead reckoning[" + id + "] No movement detected"); }
            } 
            else if(direction == 1){ // previous move was right
              if(DEADRECK_DEBUG == true) { System.out.println("Dead reckoning[" + id + "] Moving right"); }
              entity.relocate(entity.getX()+config.speed().player().getDistance(), entity.getY()); // increase x by player move speed
            } else if (direction == 2){ // previous move was left
              if(DEADRECK_DEBUG == true) { System.out.println("Dead reckoning[" + id + "] Moving left"); }
              entity.relocate(entity.getX()-config.speed().player().getDistance(), entity.getY());
            }
          }
        }
    }
  };
  
  public GamePanel() {
    setBackground(Color.BLACK);
  setForeground(Color.BLACK);
  }

  @Override
  protected void paintComponent(Graphics graphics) {
    super.paintComponent(graphics);
    if (gameOn) {
      final GraphicalEntityVisitor painter = new PaintingVisitor(graphics,this);
      graphics.setColor(Color.WHITE);
      graphics.setFont(new Font("Courier",Font.BOLD,15));
      entitiesLock.readLock().lock();
      for (Drawable entity : entityMap.values()) {
        entity.draw(painter);
      }
      entitiesLock.readLock().unlock();
    } else {
      graphics.drawImage(centerImg,centerImgPos.getFirst(),centerImgPos.getSecond(),this);
    }
  }

  /** Prepare the panel for a new game. */
  public void init() {
    gameOn = true;
    playerAvatarNumber = 0;
    entitiesLock.writeLock().lock();
    entityMap.clear();
    entitiesLock.writeLock().unlock();
  }

  /**
   * End the game and draw an image in the center.
   *
   * <p>Game entities are being wiped out, and the image will remain in the center of the panel
   * until {@link #init() init} is called.
   *
   * @throws NullPointerException if argument is {@code null}.
   */
  public void showImage(BufferedImage centerImg) {
    if (centerImg == null) {
      throw new NullPointerException();
    }
    this.centerImg = centerImg;
    int width = centerImg.getWidth();
    int height = centerImg.getHeight();
    int posX = (config.frame().getWidth() - width) / 2;
    int posY = (config.frame().getHeight() - height) / 2;
    centerImgPos = new Couple<>(posX,posY);
    gameOn = false;
    paintComponent(getGraphics());
  }

  /**
   * Refresh all entities by checking them against a list of updates.
   *
   * <p>Entities which are not found in the list of updates are removed, and new entities which are
   * found only in the list of updates are added. The coordinates of entities are changed
   * accordingly. After this method is executed, the new set of entities is going to be equal with
   * the list of updates.
   *
   * @throws NullPointerException if argument is {@code null}.
   */
  public void refreshEntities(List<Entity> updates) {

    // move player based on their previous movement direction
    if(DEADRECK_ENABLED) { executor.scheduleAtFixedRate(deadReckon, 0, DEADRECK_DELAY, TimeUnit.MILLISECONDS); } 

    if (updates == null) {
      throw new NullPointerException();
    }
    final List<Integer> elim = new ArrayList<>();
    final boolean[] mark = new boolean[updates.size()];
    entitiesLock.readLock().lock();
    for (Map.Entry<Integer,GraphicalEntity> entry : entityMap.entrySet()) {
      boolean found = false;
      int index = 0;
      for (Entity it : updates) {
        if (entry.getKey().equals(it.getId())) {
          found = true;
          entry.getValue().relocate(it.getX(),it.getY());
          mark[index] = true;
          break;
        }
        ++index;
      }
      if (!found) {
        elim.add(entry.getKey());
      }
    }
    entitiesLock.readLock().unlock();
    
    /* Remove entities not found in the updates. */
    entitiesLock.writeLock().lock();
    for (int key : elim) {
      entityMap.remove(key);
    }

    /* Add entities not found in the map. */
    int index = 0;
    for (Entity it : updates) {
      if (!mark[index]) {
        GraphicalEntity spawned = factory.create(it);
        entityMap.put(it.getId(),spawned);
      }
      ++index;
    }
    entitiesLock.writeLock().unlock();
  }

  /**
   * @param id player ID.
   * @param name player name.
   *
   * @throws NullPointerException if an argument is {@code null} of if the {@code id} could not
   *     be found.
   */
  public void setPlayer(int id, String name) {
    if (name == null) {
      throw new NullPointerException();
    }
    final Player player = (Player) entityMap.get(id);
    if (player == null) {
      throw new NullPointerException();
    }
    player.setName(name);
    player.setAvatarNumber(playerAvatarNumber++);
  }

  /**
   * Change the spatial coordinates of an entity.
   *
   * @param id entity ID.
   * @param newX new coordinate on x-axis.
   * @param newY new coordinate on y-axis.
   *
   * @throws NullPointerException if {@code id} could not be found.
   */
  public void relocateEntity(int id, int newX, int newY) { // client side dead reckoning here
    final GraphicalEntity entity = entityMap.get(id);
    if (entity == null) {
      throw new NullPointerException();
    }
    if(prevEntityPosition.containsKey(id)){ 
      Integer direction;
      if(prevEntityPosition.get(id) - newX < -1){ direction = 1; } // 0 = player not moving, 1 = moving right, 2 = moving left
      else if (prevEntityPosition.get(id) - newX > 1 ){ direction = 2; }
      else { direction = 0; } 
      prevEntityDirection.put(id, direction); // store the last direction of the the entity
    }
    entity.relocate(newX,newY);
    prevEntityPosition.put(id, newX); // store the last position
  }

  /**
   * Create a new entity.
   *
   * <p>If an entity already exists with the {@code id}, it is overwritten.
   *
   * @param id entity ID.
   * @param type type of the entity.
   * @param posX coordinate on x-axis.
   * @param posY coordinate on y-axis.
   */
  public void spawnEntity(int id, EntityEnum type, int posX, int posY) {
    entitiesLock.writeLock().lock();
    entityMap.put(id,factory.create(new Entity(type,id,posX,posY)));
    entitiesLock.writeLock().unlock();
    if (type.equals(PLAYER_BULLET)) {
      sound.shooting();
    }
  }

  /**
   * Remove an entity.
   *
   * <p>The entity is not going to be displayed any more, after the next repaint.
   *
   * @param id ID of the entity to be removed.
   *
   * @throws NullPointerException if the {@code id} could not be found.
   */
  public void wipeOutEntity(int id) {
    entitiesLock.writeLock().lock();
    GraphicalEntity entity = entityMap.remove(id);
    entitiesLock.writeLock().unlock();
    if (entity == null) {
      throw new NullPointerException();
    }
    if (entity.getType().equals(INVADER)) {
      sound.deadInvader();
    } else if (entity.getType().equals(PLAYER)) {
      sound.deadPlayer();
    }
  }

  /**
   * Translate an entire group of entities, all of the same {@code type}.
   *
   * @param type type of all entities in the group.
   * @param offsetX offset on X Axis.
   * @param offsetY offset on Y Axis.
   */
  public void translateGroup(EntityEnum type, int offsetX, int offsetY) {
    entitiesLock.readLock().lock();
    for (GraphicalEntity entity : entityMap.values()) {
      if (entity.getType().equals(type)) {
        entity.translate(offsetX,offsetY);
      }
    }
    entitiesLock.readLock().unlock();
  }
}
